### PR TITLE
fixing up some threading cases

### PR DESCRIPTION
**Pull request by jmarolf**
`_tasksService.LoadedProject()` assumes that all logic run inside the using is correct and never causes a deadlock.  `_tasksService.LoadedProjectAsync` explicitly uses JTF to ensure no deadlocks occur.

`GetService` blocks the calling thread where as `GetServiceAsync` will try several techniques to acquire the service in a non-blocking way before falling back to a blocking approach.

`AddSubscriptionsAsync` was locking but then calling asynchronus methods and not waiting for them to complete.  This could have led to races or interactions with the project after it had closed.

@davkean to double check my understanding of this threading code

Ask mode template forthcoming

### CODE REVIEW

**Comment by jmarolf** [comment]
@dotnet/project-system for review

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ internal abstract class CrossTargetSubscriptionHostBase : OnceInitializedOnceDis
 
         protected async Task AddInitialSubscriptionsAsync()
         {
-            using (_tasksService.LoadedProject())
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: So previously the object returned by `LoadedProject` would have been disposed at the end of the block. Do we need to do something similar to the object returned by `LoadedProjectAsync`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs [hunk 2]
```diff
@@ -249,7 +249,7 @@ await subscriber.Value.InitializeSubscriberAsync(this, _activeConfiguredProjectS
 
             await _commonServices.ThreadingService.SwitchToUIThread();
 
-            using (_tasksService.LoadedProject())
+            await _tasksService.LoadedProjectAsync(() =>
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: `ConfigureAwait(true)` for this `await`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs (hunk 1 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: No `_tasksService.LoadedProject()` blocked until it got a lock, releasing the lock when it was disposed.  `_tasksService.LoadedProjectAsync` asynchronously waits for a lock, executes the lambda, and then released the lock.

---
In reply to: [121492671](https://github.com/dotnet/project-system/pull/2419#discussion_r121492671) [](ancestors = 121492671)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs (hunk 2 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: [LoadedProjectAsync](http://index/?rightProject=Microsoft.VisualStudio.ProjectSystem&file=Utilities%5cCommonProjectSystemTools.cs&line=206) returns a JoinableTask which always resumes onto the same thread as its caller.

---
In reply to: [121514747](https://github.com/dotnet/project-system/pull/2419#discussion_r121514747) [](ancestors = 121514747)

**Comment by davkean** [comment]
Can we split this into 3 separate commits?

**Comment by jmarolf** [comment]
@davkean split this into separate commits.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs [hunk 3]
```diff
@@ -262,10 +262,12 @@ await subscriber.Value.InitializeSubscriberAsync(this, _activeConfiguredProjectS
 
                     foreach (var subscriber in Subscribers)
                     {
-                        subscriber.Value.AddSubscriptionsAsync(newProjectContext);
+                        Shell.ThreadHelper.JoinableTaskFactory.Run(()=> subscriber.Value.AddSubscriptionsAsync(newProjectContext));
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: Its unclear to me what the correct behavior is here.  This section of code is in a lock yet previously we would call `subscriber.Value.AddSubscriptionsAsync(newProjectContext)` and not observe the returned task or wait for it to complete. That seems to imply that we don't care about the lock, since we previously exited the lock statement without any guarantee that the async tasks had finished.  Here I am making the assumption that this was an error and we only want to exit the lock once all subscriptions have been added.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs [hunk 4]
```diff
@@ -262,10 +262,12 @@ await subscriber.Value.InitializeSubscriberAsync(this, _activeConfiguredProjectS
 
                     foreach (var subscriber in Subscribers)
                     {
-                        subscriber.Value.AddSubscriptionsAsync(newProjectContext);
+                        Shell.ThreadHelper.JoinableTaskFactory.Run(() => subscriber.Value.AddSubscriptionsAsync(newProjectContext));
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: What problem are we solving here? You're adding a synchronous blocking call in a threadpool thread, which is the cause of all these issues here: https://github.com/dotnet/project-system/issues/2297.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs [hunk 5]
```diff
@@ -266,14 +265,16 @@ private JoinableTask ExecuteWithinLockAsync(Func<Task> task)
 
                     _projectConfigurationsWithSubscriptions.Add(configuredProject.ProjectConfiguration);
                 }
-            }
+
+                return Task.CompletedTask;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs]: Why is this needed?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs (hunk 4 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: Oh we're not observing it - why isn't this just an await?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs (hunk 5 above)
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs]: Ah, the differ confused me - I see it's for func, not the parent method

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs (hunk 4 above)
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: you cannot await inside a lock in C# (for obvious reasons).

Its unclear to me what the correct thing to do here is.  Previously, we were in a lock but we did not wait for async operations to complete before leaving the lock.

either:
1. This code does not need to be in the lock, in which case we should move it out of the lock
2. This code does need to be in the lock, in which case we need to actually wait for the async operations to complete or we risk bugs from threading behavior.

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: Which is why we're using semaphores elsewhere. Is this actually fixing a known race? if not, I'd leave it until I rewrite this logic of coordination between configured projects. CPS want us to approach it a different way.

Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: > Is this actually fixing a known race

I am not aware of one,.  Sounds like this is 2 and we should not exit the lock until this subscription is finished.

> Which is why we're using semaphores elsewhere

Can you show me where else in the code we do that? 

> I'd leave it until I rewrite this logic of coordination between configured projects

Fair enough, do we have an issue # for how CPS wants us to do this?

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: See _gate in LanguageServiceHost.

Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: Looks a lot better, any objections to just doing that?

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: I have objectiins to doing in 15.3 unless we're solving an actual existing bug. This becomes easy to introduce other issues and other bugs such as deadlocks on shutdown.

Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/CrossTargetSubscriptionHostBase.cs]: ie We've regressed so many times around this code - i'd prefer not to change it unless it's fixing a bug that we're running into right now.

**Comment by srivatsn** [comment]
If this is not addressing any known perf regressions then let's retarget this for 15.6

**Comment by jmarolf** [comment]
@davkean can you comment on whether you want the remaining changes or not?

**Comment by davkean** [comment]
You pulled the semaphore? I'm okay with the other changes.

**Comment by jmarolf** [comment]
@davkean yeah, the semaphore code definitely has problems but lets tackle that in a different PR.
